staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go

```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
 	StorageVersionHash string `json:"storageVersionHash,omitempty" protobuf:"bytes,10,opt,name=storageVersionHash"`
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Discovery API Types
+// APIGroupList is a list of APIGroup, to allow clients to discover the API at
+// /apis.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type DiscoveryAPIGroupList struct {
+	TypeMeta `json:",inline"`
+	// groups is a list of APIGroup.
+	// +listType=map
+	// +listMapKey=name
+	Groups []DiscoveryAPIGroup `json:"groups" protobuf:"bytes,1,rep,name=groups"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroup contains the name, the supported versions, and the preferred version
+// of a group.
+type DiscoveryAPIGroup struct {
```

Comment:
Can we structure this with ObjectMeta like a "normal" kind?

Reply:
Do we need ObjectMeta for this kind of API? It's not a resource belonging to a cluster and just serves as an API format for the /discovery endpoint. Things part of the ObjectMeta definition like ResourceVersion, CreationTimestamp, ManagedFields, etc don't really apply to this type.

Our current apis for handling `/apis` and `/apis/<group>/<version>` also only inherit TypeMeta (https://github.com/kubernetes/kubernetes/blob/6986edae7265a1de536a54ae7ba57767455896b9/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1028-L1034)

Reply:
Which fields would be set? Part of the goal is to keep the resulting page tiny (so that it's OK for everything to be in one document).

Reply:
> Which fields would be set?

ObjectMeta.Name instead of having the bare `name`.